// react-native-chat-ui/src/components/Chat.js
// flow
import React, { Component } from "react";
import { FlatList, SafeAreaView, KeyboardAvoidingView } from "react-native";
import Composer from "./Composer";
import Bubble from "./Bubble";
import Header from "./Header";
import { User, Chat, Message } from "./types";

type State = {
  messages: Array<Message>,
};

type Props = {
  user: User,
  chat: Chat,
  messages: Array<Message>,
};

class ChatUI extends Component<Props, State> {
  state = {
    localMessages: [],
  };

  onMessageSend = m =>
    this.setState(({ localMessages }) => ({ localMessages: [m, ...localMessages] }));

  render() {
    const { user, chat, messages } = this.props;
    const { localMessages } = this.state;
    const messageList = [...localMessages, ...messages];

    return (
      <KeyboardAvoidingView style={{ flex: 1 }} behavior="padding">
        <SafeAreaView style={{ flex: 1 }}>
          <Header />
          <FlatList
            data={messageList}
            style={{
              backgroundColor: "red",
            }}
            contentContainerStyle={{
              backgroundColor: "green",
            }}
            keyExtractor={item => item.id.toString()}
            renderItem={({ item }) => <Bubble message={item} />}
            inverted
          />
          <Composer onSend={this.onMessageSend} user={user} />
        </SafeAreaView>
      </KeyboardAvoidingView>
    );
  }
}

export default ChatUI;
