// react-native-chat-ui/src/components/Composer.js
import React, { Component } from "react";
import { Text, View, TextInput, TouchableOpacity } from "react-native";
import uuid from "uuid/v1";
import { User } from "./types";

type State = {
  text: string,
};
type Props = {
  onSend: Function,
  user: User,
};

class Composer extends Component<Props, State> {
  state = {
    text: "",
  };

  onChangeText = text => this.setState({ text });

  send = () => {
    const { onSend, user } = this.props;
    const { text } = this.state;

    if (text) {
      onSend({
        id: uuid(),
        sender: user,
        text,
        date: Date.now(),
      });

      this.setState({
        text: "",
      });
    }
  };

  render() {
    const { text } = this.state;

    return (
      <View
        style={{
          flexDirection: "row",
          padding: 5,
        }}
      >
        <TextInput
          style={{
            flex: 1,
            borderWidth: 1,
            paddingTop: 5,
            paddingBottom: 5,
            paddingLeft: 5,
            paddingRight: 5,
            fontSize: 16,
            maxHeight: 100,
          }}
          multiline
          placeholder="Type..."
          value={text}
          onChangeText={this.onChangeText}
        />
        <View
          style={{
            justifyContent: "flex-end",
          }}
        >
          <TouchableOpacity onPress={this.send}>
            <Text>Send</Text>
          </TouchableOpacity>
        </View>
      </View>
    );
  }
}

export default Composer;
